text/2017-10-06-api-layers.md

```diff
@@ -0,0 +1,211 @@
+# Summary
+
+This RFC introduces three layers of API for supporting diversified use cases.
+
+
+# Motivation
+
+As the Crossbeam project is growing, users want more diversified API than the existing one.
+
+First, majority of users will be happy with the current API, which interacts with the default
+singleton garbage collector.
+
+Second, some users may want to create their own garbage collector, and use the collector via a
+*handle* to it. Since the current API provides only one garbage collector, Crossbeam currently does
+not support this use case.
+
+Lastly, some users want to embed a garbage collector in another systems library, e.g. memory
+allocator and thread management environment.  Note that it is often a part of the standard library,
+thus this use case requires (at least a part of) Crossbeam to be implemented in the `#[no_std]`
+environment. Likewise, Crossbeam currently does not support this use case.
+
+In order to support these diversified use cases, this author proposes to generalize the API by
+introducing an API layer for each use case: (1) **the default garbage collector API** for the first
+use case; (2) **`Collector` API** for the second one; and (3) **internal API** for the third one.
+
+
+
+# Detailed design
+
+FYI, this RFC is fully implemented in [this
+branch](https://github.com/jeehoonkang/crossbeam-epoch/tree/handle). You can find a discussion on
+this branch in [this PR](https://github.com/crossbeam-rs/crossbeam-epoch/pull/21).
+
+
+## Internal API
+
+This is the lowest-level API to Crossbeam. First, we define a public struct `Global` and
+`Local` that contain the global and local data for garbage collection:
+
+```rust
+struct Global {
+  ...
+}
+
+struct Local {
+  ...
+}
+```
+
+And implement the following methods:
+
+```rust
+impl Global {
+    pub fn new() -> Self;
+}
+
+impl Local {
+    fn new(global: &Global) -> Self;
+
+    // `global` should be the one used to create the `Local`.
+    pub unsafe fn pin<F, R>(&self, global: &Global, f: F)
+    where F: FnOnce(&Scope) -> R;
+    // fun f with a `Scope`
+
+    // Once a `Local` is unregistered, `pin()` or `unregister()` should not be called later.
+    pub unsafe fn unregister(&self, global: &Global);
```

Comment:
Is the idea that this is called when you're not going to use the `Local` again? So it's basically a manual `drop` for `Local`?

Reply:
Yes.